File: virbius-gateway/lib/list_redis.lua
-- Redis ZSET list match (user_id / device_id / var) with TTL match-result cache.

local _M = {}

local CACHE_TTL = tonumber(os.getenv("VIRBIUS_LIST_MATCH_CACHE_TTL_SEC")) or 60
local REDIS_TIMEOUT = tonumber(os.getenv("VIRBIUS_LIST_REDIS_TIMEOUT_MS")) or 50
local match_cache = {}

local function cache_get(key, now)
    local e = match_cache[key]
    if not e then
        return nil
    end
    if now - e.cached_at > CACHE_TTL then
        match_cache[key] = nil
        return nil
    end
    if e.hit and e.score and e.score > 0 and e.score <= now then
        match_cache[key] = nil
        return nil
    end
    return e.hit
end

local function cache_put(key, hit, score, now)
    match_cache[key] = { hit = hit, score = score or 0, cached_at = now }
end

local function redis_zscore(redis_key, member)
    local redis = require("resty.redis")
    local red = redis:new()
    red:set_timeout(REDIS_TIMEOUT)
    local redis_url = os.getenv("VIRBIUS_REDIS_URL") or "127.0.0.1:6379"
    local host, port = redis_url:match("^redis://([^:/]+):?(%d*)")
    if not host then
        host, port = redis_url:match("^([^:/]+):?(%d*)")
    end
    host = host or "127.0.0.1"
    port = tonumber(port) or 6379
    local ok, err = red:connect(host, port)
    if not ok then
        ngx.log(ngx.WARN, "virbius list redis connect failed: ", err)
        return nil
    end
    local score, rerr = red:zscore(redis_key, member)
    red:set_keepalive(10000, 100)
    if score == ngx.null or score == nil then
        return nil
    end
    if rerr then
        ngx.log(ngx.WARN, "virbius list redis zscore failed: ", rerr)
        return nil
    end
    return tonumber(score)
end

local function is_active(score, now)
    if score == nil then
        return false
    end
    if score == 0 then
        return true
    end
    return score > now
end

function _M.match(redis_key, tenant_id, list_name, lookup_value)
    if not lookup_value or lookup_value == "" or not redis_key or redis_key == "" then
        return false
    end
    local now = ngx.time()
    local cache_key = (tenant_id or "") .. ":" .. (list_name or "") .. ":" .. lookup_value
    local cached = cache_get(cache_key, now)
    if cached ~= nil then
        return cached
    end
    local score = redis_zscore(redis_key, lookup_value)
    if score == nil then
        cache_put(cache_key, false, 0, now)
        return false
    end
    local hit = is_active(score, now)
    cache_put(cache_key, hit, score, now)
    return hit
end

return _M

File: virbius-gateway-agent/src/list_redis.rs
use redis::Commands;
use serde::Deserialize;
use std::collections::HashMap;
use std::env;
use std::sync::{Mutex, OnceLock};
use std::time::{Duration, Instant};

static REDIS_CLIENT: OnceLock<Option<redis::Client>> = OnceLock::new();
static MATCH_CACHE: OnceLock<Mutex<HashMap<String, CacheEntry>>> = OnceLock::new();

#[derive(Clone, Copy)]
struct CacheEntry {
    hit: bool,
    score: f64,
    cached_at: Instant,
}

#[derive(Debug, Clone, Default, Deserialize)]
pub struct RedisListIndexBlock {
    #[serde(default)]
    pub list_name: String,
    #[serde(default)]
    pub dimension: String,
    #[serde(default)]
    pub redis_key: String,
}

fn redis_client() -> Option<&'static redis::Client> {
    REDIS_CLIENT
        .get_or_init(|| {
            env::var("VIRBIUS_REDIS_URL")
                .ok()
                .filter(|s| !s.is_empty())
                .and_then(|url| redis::Client::open(url).ok())
        })
        .as_ref()
}

fn cache_ttl() -> Duration {
    let sec = env::var("VIRBIUS_LIST_MATCH_CACHE_TTL_SEC")
        .ok()
        .and_then(|s| s.parse().ok())
        .unwrap_or(60);
    Duration::from_secs(sec.max(1))
}

fn match_cache() -> &'static Mutex<HashMap<String, CacheEntry>> {
    MATCH_CACHE.get_or_init(|| Mutex::new(HashMap::new()))
}

fn is_active(score: f64, now_sec: i64) -> bool {
    if score == 0.0 {
        return true;
    }
    score > now_sec as f64
}

pub fn redis_list_match(tenant_id: &str, list_name: &str, redis_key: &str, lookup_value: &str) -> bool {
    if lookup_value.is_empty() || redis_key.is_empty() {
        return false;
    }
    let cache_key = format!("{tenant_id}:{list_name}:{lookup_value}");
    let now = Instant::now();
    let now_sec = chrono::Utc::now().timestamp();
    if let Ok(guard) = match_cache().lock() {
        if let Some(e) = guard.get(&cache_key) {
            if now.duration_since(e.cached_at) <= cache_ttl() {
                if !(e.hit && e.score > 0.0 && e.score <= now_sec as f64) {
                    return e.hit;
                }
            }
        }
    }
    let Some(client) = redis_client() else {
        return false;
    };
    let score: Option<f64> = client
        .get_connection()
        .ok()
        .and_then(|mut conn| conn.zscore(redis_key, lookup_value).ok())
        .flatten();
    let hit = score.map(|s| is_active(s, now_sec)).unwrap_or(false);
    if let Ok(mut guard) = match_cache().lock() {
        guard.insert(
            cache_key,
            CacheEntry {
                hit,
                score: score.unwrap_or(0.0),
                cached_at: now,
            },
        );
    }
    hit
}

pub fn match_redis_list_by_name(
    tenant_id: &str,
    index: &[RedisListIndexBlock],
    list_name: &str,
    explicit_value: Option<&str>,
    content: &str,
    user_id: Option<&str>,
    device_id: Option<&str>,
    client_ip: Option<&str>,
    vars: &HashMap<String, String>,
) -> bool {
    let Some(block) = index.iter().find(|b| b.list_name == list_name) else {
        return false;
    };
    if block.redis_key.is_empty() {
        return false;
    }
    let dim = block.dimension.as_str();
    if let Some(v) = explicit_value {
        return redis_list_match(tenant_id, list_name, &block.redis_key, v);
    }
    if let Some(logical) = var_logical_name(dim) {
        if let Some(val) = vars.get(logical) {
            if !val.is_empty() {
                return redis_list_match(tenant_id, list_name, &block.redis_key, val);
            }
        }
        return false;
    }
    let lookup = match dim {
        "user_id" => user_id,
        "device_id" => device_id,
        "ip" | "ip_cidr" => client_ip,
        "keyword" | "content" => Some(content),
        _ => None,
    };
    lookup
        .filter(|v| !v.is_empty())
        .map(|v| redis_list_match(tenant_id, list_name, &block.redis_key, v))
        .unwrap_or(false)
}

fn var_logical_name(dim: &str) -> Option<&str> {
    dim.strip_prefix("var:").filter(|s| !s.is_empty())
}
